source/extensions/compression/zstd/common/dictionary_manager.h

```diff
@@ -0,0 +1,127 @@
+      const auto data = Config::DataSource::read(source, false, api);
+      auto dictionary = DictionaryPtr(builder_(data.data(), data.length()));
+      auto id = getDictId(dictionary.get());
+      // If id == 0, the dictionary is not conformant to Zstandard specification, or empty.
+      if (id == 0) {
+        throw EnvoyException("Illegal Zstandard dictionary");
+      }
+      dictionary_map_->emplace(id, std::move(dictionary));
+      if (source.specifier_case() ==
+          envoy::config::core::v3::DataSource::SpecifierCase::kFilename) {
+        has_watched = true;
+        const auto& filename = source.filename();
+        watcher_->addWatch(filename, Filesystem::Watcher::Events::Modified |
+                                         Filesystem::Watcher::Events::MovedTo,
+                           [this, id, filename](uint32_t) { onDictionaryUpdate(id, filename); });
+      }
+    }
+
+    if (!has_watched) {
+      watcher_.reset();
+    }
+  };
+
+  T* getDictionary(bool want_first, unsigned id) {
+    DictionaryMapPtr dictionary_map;
+    {
+      if (watcher_) {
+        absl::ReaderMutexLock l(&mutex_);
+      }
+      dictionary_map = dictionary_map_;
+    }
+
+    if (dictionary_map) {
+      typename absl::flat_hash_map<unsigned, DictionaryPtr>::iterator it;
+      if (want_first) {
+        it = dictionary_map->begin();
+      } else {
+        it = dictionary_map->find(id);
+      }
+      if (it != dictionary_map->end()) {
+        return it->second.get();
+      }
+    }
+
+    return nullptr;
+  };
+
+  T* getDictionaryById(unsigned id) { return getDictionary(false, id); };
+
+  T* getFirstDictionary() { return getDictionary(true, 0); };
+
+private:
+  class DictionaryPtr : public std::shared_ptr<T> {
+  public:
+    DictionaryPtr(T* object) : std::shared_ptr<T>(object, deleter) {}
+  };
+  using DictionaryMapPtr = std::shared_ptr<absl::flat_hash_map<unsigned, DictionaryPtr>>;
+
+  void onDictionaryUpdate(unsigned origin_id, const std::string& filename) {
+    const auto data = api_.fileSystem().fileReadToEnd(filename);
+    if (!data.empty()) {
+      auto dictionary = DictionaryPtr(builder_(data.data(), data.length()));
+      auto id = getDictId(dictionary.get());
+      // Keep origin dictionary if the new is illegal
+      if (id != 0) {
+        auto dictionary_map = std::make_shared<absl::flat_hash_map<unsigned, DictionaryPtr>>();
+        dictionary_map->reserve(dictionary_map_->size());
+        for (auto it = dictionary_map_->begin(); it != dictionary_map_->end(); ++it) {
+          // Not replace mode or not origin id, keep it
+          if (!replace_mode_ || it->first != origin_id) {
+            dictionary_map->emplace(it->first, it->second);
+          }
+        }
+        dictionary_map->emplace(id, std::move(dictionary));
+
+        {
+          absl::WriterMutexLock l(&mutex_);
+          dictionary_map_ = std::move(dictionary_map);
+        }
+      }
+    }
+  }
+
+  absl::Mutex mutex_;
```

Comment:
This will be frowned upon definitely. Could you avoid locking with the thread_local API (envoy/thread_local/thread_local.h)?

Reply:
Hi @rojkov , thanks for you review! I'm ported how the [ssl_ctx_](https://github.com/rainingmaster/envoy/blob/main/source/extensions/transport_sockets/tls/ssl_socket.cc#L395) module updates files and sync to various threads before. But you are correct, TLS must more efficient than lock. I have update it now